### PR TITLE
moss: upstreamability refactor

### PR DESCRIPTION
**Pull request by jleightcap**
This is a collection of our work (@albertchae, @DMills27, @jasonodoom, @Chickensoupwithrice) to refactor some of the definitions in `flake.nix` towards matching conventions of upstream nixpkgs.

Including:

- leverage upstream packages with overrides rather than re-defining each,
- `callPackage` idiom,
- destructuring derivations into individual files

### CODE REVIEW

**Comment by albertchae** [comment]
@hansfbaier we need to make some changes on this PR to catch up to your latest modifications, but we had some questions working on the upstreamability refactor that we put into a temporary PR in our fork https://github.com/ngi-nix/toolchain-nix/pull/1, if you could help us understand better. Thanks!

File: flake.nix [hunk 1]
```diff
@@ -3,396 +3,110 @@
...
-          buildInputs = [ readline ];
+            src = fetchFromGitHub {
+              owner = "yosyshq";
+              repo = "yosys";
+              rev = "${prev.pname}-${version}";
+              hash = "sha256-IjT+G3figWe32tTkIzv/RFjy0GBaNaZMQ1GA8mRHkio=";
+            };
 
-          installPhase = "mkdir -p $out/bin && mv abc $out/bin";
+            doCheck = true; # FIXME(ac): can we turn these back on?
```
Comment (hansfbaier) [inline flake.nix:65]: Yes, I think so. Those tests just took a long time and hindered the development process, so I turned them off.

**Comment by hansfbaier** [comment]
@jleightcap Unfortunately there are no plans to move to a newer yosys, because there are more important priorities and also because newer versions seem to crash more often, especially when using abc9.

File: flake.nix [hunk 2]
```diff
@@ -24,375 +23,76 @@
...
-              license = licenses.gpl2;
-            };
-          };
+      # Provide some binary packages for selected system types.
+      packages = forAllSystems (system:
+        let
+          pkgs = nixpkgsFor.${system};
+          inherit (pkgs) lib callPackage stdenv fetchgit fetchFromGitHub;
+        in rec {
+          default = yosys;
```
Comment (hansfbaier) [inline flake.nix:35]: Having yosys as the default package, does not make much sense to me. I chose that in lack of something better.
What would you suggest?

**Comment by hansfbaier** [comment]
@jleightcap Yes, help with upstreaming would be greatly appreciated, because working on the toolchain itself is already very challenging.

**Pull request merged**